File: compiler/scheme2/lexer.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
#
# Copyright (c) 2015 ASMlover. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list ofconditions and the following disclaimer.
#
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materialsprovided with the
#    distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import error
import token

class Lex(object):
    def __init__(self, scan):
        self.scan = scan
        self.list = []
        self.paren_depth = 0

    def lex(self):
        scan = self.scan
        self.list = []
        while scan.next():
            pos = scan.get_pos()
            if scan.get() == '(':
                self.list.append(token.Token('bparen', scan.emit(), pos))
                self.paren_depth += 1
                self.lex_list()
            elif scan.get() == ')':
                self.list.append(token.Token('eparen', scan.emit(), pos))
                self.paren_depth -= 1
                if self.paren_depth > 0:
                    self.lex_list()
                elif self.paren_depth < 0:
                    self.paren_depth += 1
                    raise error.Error('too many end parens').error()
                else:
                    return self.list
            elif scan.get() == '#':
                while scan.get() != '\n':
                    scan.next()
                scan.emit()
            else:
                dump = scan.emit()
                if len(dump.strip()) > 0:
                    error.Error('unknown \'' + (dump if len(dump) <= 10 else dump[0:10-3] + '...') + '\'', pos=scan.get_pos()).error()
        return self.list

    def lex_list(self):
        scan = self.scan
        self.lex_space()
        while True:
            if not self.lex_id():
                if not self.lex_number():
                    if not self.lex_str():
                        break
            self.lex_space()
    
    def lex_space(self):
        scan = self.scan
        while scan.next():
            if not scan.get().isspace():
                scan.backup()
                break
        scan.emit()

    def lex_id(self):
        scan = self.scan
        pos = scan.get_pos()
        if not scan.next():
            return False
        if not scan.get().isalpha() and not self.is_symbol(scan.get()):
            scan.backup()
            return False
        while scan.next():
            if not scan.get().isalnum() and not self.is_symbol(scan.get()):
                scan.backup()
                break
        self.list.append(token.Token('id', scan.emit(), pos))
        return True

    def lex_number(self):
        scan = self.scan
        pos = scan.get_pos()
        while scan.next():
            if not scan.get().isdigit():
                scan.backup()
                break
        if scan.len() == 0:
            return False
        self.list.append(token.Token('number', scan.emit(), pos))
        return True

    def lex_str(self):
        scan = self.scan
        pos = scan.get_pos()
        scan.next()
        if scan.get() == '"':
            while scan.next():
                if scan.get() == '"':
                    break
        else:
            scan.backup()
            return False
        self.list.append(token.Token('str', scan.emit()[1:-1], pos))
        return True

    def is_symbol(self, c):
        return c in [
            ':',
            '!',
            '`',
            '~',
            '@',
            '#',
            '$',
            '%',
            '^',
            '&',
            '+', 
            '-',
            '*',
            '/',
            '=',
            '>',
            '<',
            '{',
            '}',
            '[',
            ']',
            '.',
        ]

File: c/tadpole/lex/lexer.c
/*
 * Copyright (c) 2024 ASMlover. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list ofconditions and the following disclaimer.
 *
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materialsprovided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <ctype.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <lex/token.h>
#include <lex/lexer.h>

typedef struct Lexer {
  const char*                     start;
  const char*                     current;
  int                             lineno;
} Lexer;

#define LEXER_LITERIAL_LEN()      (sz_t)(lexer->current - lexer->start)

static inline bool is_alpha(char c) {
  return isalpha(c) || c == '_';
}

static inline bool is_alnum(char c) {
  return isalnum(c) || c == '_';
}

static inline bool is_digit(char c) {
  return isdigit(c);
}

static inline bool is_at_end(Lexer* lexer) {
  return 0 == *lexer->current;
}

static inline char advance(Lexer* lexer) {
  return *lexer->current++;
}

static inline char peek(Lexer* lexer) {
  return *lexer->current;
}

static inline char peek_next(Lexer* lexer) {
  if (is_at_end(lexer))
    return 0;
  return lexer->current[1];
}

static inline bool match(Lexer* lexer, char expected) {
  if (is_at_end(lexer))
    return false;

  if (*lexer->current == expected) {
    ++lexer->current;
    return true;
  }
  return false;
}

static void skip_whitespace(Lexer* lexer) {
  for (;;) {
    char c = peek(lexer);
    switch (c) {
    case ' ': case '\r': case '\t': advance(lexer); break;
    case '\n': ++lexer->lineno; advance(lexer); break;
    case '/':
      if ('/' == peek_next(lexer)) {
        while ('\n' != peek(lexer) && !is_at_end(lexer))
          advance(lexer);
      }
      else {
        return;
      }
      break;
    default: return;
    }
  }
}

static Token identifier(Lexer* lexer) {
  while (is_alpha(peek(lexer)) || is_digit(peek(lexer)))
    advance(lexer);

  const char* literal = lexer->start;
  return make_token(get_keyword_kind(literal), literal, LEXER_LITERIAL_LEN(), lexer->lineno);
}

static Token number(Lexer* lexer) {
  while (is_digit(peek(lexer)))
    advance(lexer);

  if ('.' == peek(lexer) && is_digit(peek_next(lexer))) {
    advance(lexer);

    while (is_digit(peek(lexer)))
      advance(lexer);
  }

  return make_token(TOKEN_NUMBER, lexer->start, LEXER_LITERIAL_LEN(), lexer->lineno);
}

static Token string(Lexer* lexer) {
  while ('"' == peek(lexer) && !is_at_end(lexer)) {
    if ('\n' == peek(lexer))
      ++lexer->lineno;

    advance(lexer);
  }

  int lineno = lexer->lineno;
  if (is_at_end(lexer))
    return make_error_token("Unterminated string.", lineno);

  advance(lexer);
  return make_token(TOKEN_STRING, lexer->start, LEXER_LITERIAL_LEN(), lineno);
}

Lexer* lexer_init(const char* source_code) {
  Lexer* lexer = (Lexer*)malloc(sizeof(Lexer));
  if (NULL != lexer) {
    lexer->start = source_code;
    lexer->current = source_code;
    lexer->lineno = 1;
  }
  return lexer;
}

void lexer_destroy(Lexer* lexer) {
  if (NULL != lexer)
    free(lexer);
}

Token lexer_next_token(Lexer* lexer) {
  return make_error_token("Error", 0);
}
